domain/application/state/application.go

```diff
@@ -1874,3 +1874,78 @@ func decodeArchitecture(arch int) (application.Architecture, error) {
 		return -1, fmt.Errorf("unsupported architecture: %d", arch)
 	}
 }
+
+// GetApplicationIDByUnitName returns the application ID for the named unit,
+// returning an error satisfying [applicationerrors.UnitNotFound] if the unit
+// doesn't exist.
+func (st *ApplicationState) GetApplicationIDByUnitName(
+	ctx context.Context,
+	name coreunit.Name,
+) (coreapplication.ID, error) {
+	db, err := st.DB()
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+
+	queryUnit := `
+SELECT application_uuid AS &applicationID.uuid
+FROM unit
+WHERE name = $unitNameAndUUID.name
+`
+	query, err := st.Prepare(queryUnit, applicationID{}, unitNameAndUUID{})
+	if err != nil {
+		return "", internalerrors.Errorf("preparing query for unit %q: %w", name, err)
+	}
+
+	unit := unitNameAndUUID{Name: name}
+	var app applicationID
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, query, unit).Get(&app)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return applicationerrors.UnitNotFound
+		}
+		return err
+	})
+	if err != nil {
+		return "", internalerrors.Errorf("querying unit %q application id: %w", name, err)
+	}
+	return app.ID, nil
+}
+
+// GetCharmModifiedVersion looks up the charm modified version of the given
+// application. Returns [applicationerrors.ApplicationNotFound] if the
+// application is not found.
+func (st *ApplicationState) GetCharmModifiedVersion(ctx context.Context, id coreapplication.ID) (int, error) {
+	db, err := st.DB()
+	if err != nil {
+		return -1, errors.Trace(err)
+	}
+
+	type cmv struct {
```

Comment:
Why not have this in the types.go file?

Reply:
my two cents (i ask myself the same question :) ):

For me it is ok to not introduce a type in types.go as long as it is used only once. Cluttering types.go isn't a good thing, and i find this version very readable, personnaly (I can have the whole code logic in my screen, make me happy)

Reply:
Because it was only used in one place. If we use it more than once, it can be moved.